packages/path_provider/path_provider_windows/lib/path_provider_windows.dart

```diff
@@ -0,0 +1,93 @@
+// Copyright 2020 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:ffi';
+
+import 'package:ffi/ffi.dart';
+import 'package:win32/win32.dart';
+import 'package:path_provider_platform_interface/path_provider_platform_interface.dart';
+import 'package:path_provider_windows/folders.dart';
+
+/// The Windows implementation of [PathProviderPlatform]
+///
+/// This class implements the `package:path_provider` functionality for Windows
+class PathProviderWindows extends PathProviderPlatform {
+  /// Path to the temporary directory on the device that is not backed up and is
+  /// suitable for storing caches of downloaded files.
+  ///
+  /// On Windows, this the path specified by the TMP environment variable, or
+  /// the TEMP environment variable, or the USERPROFILE environment variable,
+  /// or the Windows directory, in order of preference. Windows does not
+  /// guarantee that the path exists or is writeable to.
+  @override
+  Future<String> getTemporaryPath() {
+    final buffer = allocate<Uint16>(count: MAX_PATH + 1).cast<Utf16>();
+    final length = GetTempPath(MAX_PATH, buffer);
+
+    if (length == 0) {
+      final error = GetLastError();
+      free(buffer);
+      throw WindowsException(error);
+    } else {
+      var path = buffer.unpackString(length);
+
+      // GetTempPath adds a trailing backslash, but SHGetKnownFolderPath does not.
+      // Strip off trailing backslash for consistency with other methods here.
+      if (path[path.length - 1] == '\\') {
+        path = path.substring(0, path.length - 1);
+      }
+      free(buffer);
+      return Future.value(path);
+    }
+  }
+
+  /// Path to a directory where the application may place application support
+  /// files.
+  @override
+  Future<String> getApplicationSupportPath() =>
+      getPath(WindowsKnownFolder.ProgramData);
```

Comment:
There's a mismatch here between what the Windows API gives and what a path_provider user is likely to expect; on mobile these folders are in the app sandbox, so it's reasonable to just create files in the directory returned by `getApplicationSupportPath`; in a Win32 app (unless running in a compatibility hypervisor, IIRC) it's the top level folder that all applications share.

For the FDE plugin what I did is append a directory using the executable name (from `GetModuleFileName`). I think we need do that (or something like it if there's a better value we have access to) there.

(We may want to add an API to path_provider_windows that allows people to manually set it to a value of their choosing, overriding the default of exe name.)

Reply:
Wouldn't it make sense to place all app data in the `userprofile/AppData/local/AppName` as default?

Reply:
Oh, I missed that difference from FDE. I was using RoamingAppData; that (or LocalAppData) seems like the right location for application support data. ProgramData sounds like it's not user-specific.